### ISSUE TITLE
Random QA tests failures

### ISSUE THREAD

**Issue by Coduz** [eclipse/kapua#600]
Hi all,

after merging #583 some random test failures started to appear more frequently. We had already some random failures on qa tests but they were occasional.

After digging a bit I discovered that is a problem related to how Liquibase script are loaded and the fact that the Kapua Liquibase Client  is run from many different places concurrently. 

The KapuaLiquibaseClient was implemented to copy resources from the jar to a tmp directory and then execute it immediately. With the improvement contained on #583  all scripts gets copied and then executed. 

For the fact that KapuaLiquibaseClient is run from different places it can happen that the copied resources gets overwritten by another KapuaLiquibaseClient is executing them, with even more chance that now that apply of Liquibase scripts takes a bit longer.

I'll fix this making execution of Liquibase script `synchronized` as a simple patch.

Regards,

_Alberto_

**Comment by muros-ct** [comment]
I think we should "re-think" Liquibase and its usage.
Because as we discussed, right now it is used mainly for running tests. And it causes issues, maybe also because it is not used "properly".

**Comment by Coduz** [comment]
Hi @muros-ct,

Yes, there is already a issue for this task. 
The first implementation with plain SQL databases is not suitable for our needs, nor for testing nor for an actual installation.
On #586 I've discussed it.

PR #606 should take us to a point that at least from test scenarios we can run them without problems.

Regards, 

_Alberto_

**Comment by ctron** [comment]
Can you describe what the problem actually is?

**Comment by ctron** [comment]
Sorry, the email I received cut off part of your analysis.

I think there are two ways to solve this concurrency issue. Use locks, as you did. Or use different directories for copying. In the end this is a simple issue of not being multi threading safe. So it should be an easy fix.

**Comment by Coduz** [comment]
@ctron  Please read #586  for the different directories.

**Comment by andrejn-ct** [comment]
Hi all!

FWIW, I noticed that in a Windows environment a lot of tests started to consistently fail. 
The reasons for the failures seem to be always the same. Something is preventing a file to be deleted (error message _java.io.IOException: Unable to delete file: C:\cygwin64\tmp\kapua-liquibase\changelog-user-master.xml_, the exact file varies with the test cases).
This issue seems to be Windows only as both on Linux and on macOS the tests pass fine. Also I can observe the exact same behavior both from Cygwin and the windows cmd.
Just building the code passes fine.
The tests that fail are:
kapua-liquibase
kapua-user-internal
kapua-security-shiro
kapua-device-registry-internal
kapua-datastore-internal
kapua-account-internal
kapua-qa

**Comment by Coduz** [comment]
@andrejn-ct I'll take a look on those. 
I'll probably work on this problem on another PR so we can stabilize our CI builds and stop to get random failures. Is it ok?

**Comment by ctron** [comment]
So just to understand it right, unit tests are executed in parallel? How do database instances get isolated in this case?

**Comment by ctron** [comment]
Please have a look at PR #610 for a lock free solution to this.

**Comment by Coduz** [comment]
Hi all, 

Please add further ideas and opinion on #586.

Thanks,

_Alberto_
